packages/react-aria-components/src/DropZone.tsx

```diff
@@ -0,0 +1,132 @@
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {AriaLabelingProps} from '@react-types/shared';
+import {ContextValue, Provider, RenderProps, SlotProps, useContextProps, useRenderProps, useSlot} from './utils';
+import {DropOptions, mergeProps, useClipboard, useDrop, useFocusRing, useHover, useId, VisuallyHidden} from 'react-aria';
+import {FileTriggerContext} from './FileTrigger';
+import {filterDOMProps} from '@react-aria/utils';
+import React, {createContext, ForwardedRef, forwardRef} from 'react';
+import {TextContext} from './Text';
+import {useLabels} from '@react-aria/utils';
+
+export interface DropZoneRenderProps {
+  /**
+   * Whether the dropzone is currently hovered with a mouse.
+   * @selector [data-hovered]
+   */
+  isHovered: boolean,
+  /**
+   * Whether the dropzone is focused, either via a mouse or keyboard.
+   * @selector :focus
+   */
+  isFocused: boolean,
+  /**
+   * Whether the dropzone is keyboard focused.
+   * @selector [data-focus-visible]
+   */
+  isFocusVisible: boolean,
+  /**
+   * Whether the dropzone is the drop target.
+   * @selector [data-drop-target]
+   */
+  isDropTarget: boolean
+}
+// note: possibly add isDisabled prop in the future
+export interface DropZoneProps extends Omit<DropOptions, 'getDropOperationForPoint' | 'onInsert' | 'onRootDrop' | 'onItemDrop' | 'onReorder'>, RenderProps<DropZoneRenderProps>, SlotProps, AriaLabelingProps {}
+
+export const DropZoneContext = createContext<ContextValue<DropZoneProps, HTMLDivElement>>(null);
+
+function DropZone(props: DropZoneProps, ref: ForwardedRef<HTMLDivElement>) {
+  [props, ref] = useContextProps(props, ref, DropZoneContext);
+  let {dropProps, isDropTarget} = useDrop({...props, ref});
+  let {hoverProps, isHovered} = useHover({});
+  let {focusProps, isFocused, isFocusVisible} = useFocusRing();
+  let [fileTriggerRef, hasFileTrigger] = useSlot(); 
+
+  let textId = useId();
+  let labelProps = useLabels({'aria-labelledby': textId});
+
+  let {clipboardProps} = useClipboard({
+    onPaste: (items) => props.onDrop?.({  
+      type: 'drop',
+      items,
+      x: 0,
+      y: 0,
+      dropOperation: 'copy'
+    })
+  });
+
+  let renderProps = useRenderProps({ 
+    ...props,
+    values: {isHovered, isFocused, isFocusVisible, isDropTarget},
+    defaultClassName: 'react-aria-DropZone'
+  });
+  let DOMProps = filterDOMProps(props);
+  delete DOMProps.id;
+
+  // passed to the FileTrigger component
+  let onInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
```

Comment:
Do we need these to be tied together? Since they control both elements, it seems like a user could do something like this:

```jsx
<DropZone onDrop={onDrop}>
  <FileTrigger onChange={onChange}>
     ...
  </FileTrigger>
</DropZone>
```

Maybe slightly more work but maybe not a big deal?